### ISSUE TITLE
What’s the difference/relationship between registry and distribution/registry?

### ISSUE THREAD

**Issue by x-yuri** [distribution/distribution-library-image#91]
I've tried asking on [Docker Forums][1], but nobody seems to know the answer.

[1]: https://forums.docker.com/t/whats-the-difference-relationship-between-registry-and-distribution-registry/72158

[`distribution/registry`][2]'s page on DockerHub says:

[2]: https://hub.docker.com/r/distribution/registry

> WARNING: NOT the registry official image!!! What you LIKELY want is docker pull registry

But I can see that [`registry`][5] simply [builds][3] `distribution/registry` image, takes the binary out of it, then builds it [again][4]. What’s the deal?

[5]: https://hub.docker.com/_/registry
[3]: https://github.com/docker/distribution-library-image/blob/registry-2.7.1/update.sh
[4]: https://github.com/docker/distribution-library-image/blob/registry-2.7.1/amd64/Dockerfile

And now that I look at it, there's also [`docker-library/official-images`][6] repository. Okay, one might guess that the development is done in [`docker/distribution`][7], but the images are [released][9] by `docker-library/official-images`. But then what is [`docker/distribution-library-image`][8] repository for? Can you elaborate on the process by which `docker/distribution` becomes a package?

[6]: https://github.com/docker-library/official-images
[7]: https://github.com/docker/distribution
[8]: https://github.com/docker/distribution-library-image
[9]: https://github.com/docker-library/official-images/blob/29958f4d83cd21938c43fec7410e2e3316195e98/library/registry#L3

**Comment by tianon** [comment]
`distribution/registry` is likely a Docker Hub "Automated Build" of https://github.com/docker/distribution/blob/master/Dockerfile, so you're likely to get the latest and greatest development improvements from `master` by using that image (assuming it's still actively rebuilding for pushes; last change to GitHub was ~29 days ago, last push was ~3 months ago, so doesn't *appear* to be actively updating).

https://github.com/docker/distribution is the source repository for the "Docker Open Source Registry".

https://github.com/docker/distribution-library-image is where the packaging of *releases* of the Open Source Registry happen targeting https://github.com/docker-library/official-images (which is where all top-level unprefixed Docker Hub images are maintained as part of the "Docker Official Images Program"; see https://github.com/docker-library/official-images#what-do-you-mean-by-official).

**Comment by Jean-Baptiste-Lasselle** [comment]
> then

Hi @x-yuri , here you go with the relation (let's speak machine language, or french, I propose you machine first) : 

```bash 
# Here you are on the machine on which you build your registry:x.y.z
# In this example, the machine I build my image on is run with Cent OS 7

# So A./ you just install a package, namely [docker-distribution]
sudo yum install -y docker-distribution

# Then, guess what you got on your machine?
sudo ls -allh /bin/registry

# Yes! => https://github.com/docker/distribution-library-image/blob/b804bcd6a44de56fcf8ff3f0ababae7479ad608e/update.sh#L32

# Alright, now that we've got our {docker build dependencies}
# Yes, we can 
export URI_OF_U_NO_WAT=https://github.com/docker/distribution-library-image
mkdir ./imfrench
git clone "$URI_OF_U_NO_WAT" ./imfrench
cd ./imfrench

# And you just run the recipe, following https://github.com/docker/distribution-library-image.git 's README.md

# That's just the GO compiling env, needed by https://github.com/docker/distribution-library-image.git   's  recipe : 

export GOARCH=amd64
export GOOS=linux

# And now let's run it : 
# oh, the recipe (its a release automation recipe) obviously offers you the possibility to speciy which version you want to build)


export BRANCH_NAME=master
./update.sh $BRANCH_NAME

```

**Comment by Jean-Baptiste-Lasselle** [comment]
Update to @x-yuri , heres how, mainly I run  the whole thing, but only modified to add 2 env. variables, to make it all possible 100% offline : 

```bash 
export OPS_HOME=`mktemp -d --tmpdir bootie-dockie.XXXXXX`
export URI_REPO_BUILD_IMAGE_REGISTRY=https://github.com/docker/distribution-library-image
export URI_REPO_CODE_SOURCE_REGISTRY=https://github.com/docker/distribution-library-image
export BRANCH_NAME_OR_RELEASE_VERSION_TAG
export REGISTRY_V2_DESIRED_RELEASE_VERSION=2.1.7

# Donc en fait ce qu'il y a c'est que le script upate.Sh va faire un git clone
# docn il faut installer GOLANG, et définir les variables d'envrionnement 
export GOOS=linux 
export GOARH=amd64 

# two env. variables holding the GO VERSION. just for test purposes, not 
# to ever use in production, or merge on any branch 
export VERSION=1.12.9
export GO_VERSION=$VERSION

export LIEN_TELECHARGEMENT_TAR_GOLANG=https://dl.google.com/go/go$VERSION.$GOOS-$GOARH.tar.gz
export NOM_FICHIER_TAR_GOLANG=go$VERSION.$GOOS-$GOARH.tar.gz

# install golang
curl -L -o $NOM_FICHIER_TAR_GOLANG $LIEN_TELECHARGEMENT_TAR_GOLANG

tar -C /usr/local -xzf $NOM_FICHIER_TAR_GOLANG
export PATH=$PATH:/usr/local/go/bin

echo "Now let's test go works"

export TEMP_DIR=`mktemp -d --tmpdir gotest.XXXXXX`
mkdir -p $TEMP_DIR/src
export TEST_GO_SRC_FILE=$TEMP_DIR/src/hello.go
touch $TEST_GO_SRC_FILE
echo 'package main' >> $TEST_GO_SRC_FILE
echo ' ' >> $TEST_GO_SRC_FILE
echo 'import "fmt"' >> $TEST_GO_SRC_FILE
echo 'func main() {' >> $TEST_GO_SRC_FILE
echo '	fmt.Printf("hello, from Test Go Executable Built successfully ! \n")' >> $TEST_GO_SRC_FILE
echo '}' >> $TEST_GO_SRC_FILE
echo ' ' >> $TEST_GO_SRC_FILE


cd $TEMP_DIR/src
go build

./hello
cd $OPS_HOME
rm -fr $TEMP_DIR

echo "Right, now that we have the GOLANG environement set here, let's get to build registry from source : "

mkdir $OPS_HOME/buildfromsrc/

git clone "$URI_REPO_BUILD_IMAGE_REGISTRY" $OPS_HOME/buildfromsrc/
cd $OPS_HOME/buildfromsrc/

# et moi j'ai rajouté la variable d'envrionnement 
# export URI_REPO_CODE_SOURCE_REGISTRY, dans le script "update".

# 
# Le Script update.sh va : 
# => faire le [ git clone -b $REGISTRY_V2_DESIRED_RELEASE_VERSION $URI_REPO_CODE_SOURCE_REGISTRY ] , 
# => faire le build from source,
# => puis mettre à jour deux fichiers, dans chacun des répertoires : 
#      + $OPS_HOME/buildfromsrc/amd64 :   
#      + $OPS_HOME/buildfromsrc/arm64
#      + $OPS_HOME/buildfromsrc/arm
# 
#     Ces deux fichiers sont le fichier exécutable, résultat du build GOLANG, et le fichier [./docker-entry-point.sh]
# 



chmod +x ./update.sh
./update.sh

#
# ### Une fois cette mise à  jour des fichiers de contexte de build docker terminée
# 
#     il ne reste plus qu'à aller faire le build docker de l'image `registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION` 
#     pour chaque architecture de processeur (importance de l'ARM)
# 

echo "Ah, si, ici, il manque une chose : il faut pouvoir faire le build de l'image de base [alpine:3.8] pour les 3 arch processuers [amd64/alpine:3.8], [arm64/alpine:3.8], [arm/alpine:3.8]  " && exit 1

docker build -t amd64/registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION amd64

docker build -t arm64:registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION arm64
docker build -t arm:registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION arm

#
# ### Maintenant il faut lancer la recette de provision docker-compose de mon private docker registry 
# 
#     et ensuite je pousse les 6 images docker que j'ai du créer localement (docker push all those to ur new private registry) : 

# + amd64/alpine:3.8
# + arm64/alpine:3.8
# + arm/alpine:3.8  
# + amd64/registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION
# + arm64:registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION
# + arm:registry:$REGISTRY_V2_DESIRED_RELEASE_VERSION

# 
# ### Pour finir, je fais un backup, il faut donc avant la lancée en prod, fair eunrtest du backup / restore, pour l'exploitation complète du registry.
#     (then immefdiately backup your booted repo, to get a reproduceable first consisent initial state for production )
#      Note: meaning the very first time you do that, you WILL have to (maybe write) (surely test) your backup and restore automatic recipes
#      Only after that your fit anfd good to go for production.
# 


#
# ### Manquent : https, authentification, base de données, et une ou deux autres conneries que j'ai oubliées.
# 
# 



```

**Comment by thetredev** [comment]
@tianon I'm still confused. There is a [`distribution/distribution`](https://hub.docker.com/r/distribution/distribution) on Docker Hub, which is basically the same as [`registry`](https://hub.docker.com/_/registry), at least on first glance. However, they have different digests. Also, `distribution` is from CNCF apparently, and `registry` is from the "official images".

What I get from this: `distribution` is a CNCF project which forms the fundamental APIs to host OCI images. `registry` is somehow a "Docker specific" version of `distribution` (but I don't know what's specific about it when it comes to images).

Am I right with my assumptions? If so, do you know the "Docker specifics"? I guess it's something about OCI image specifications, but I can be very wrong.

**Comment by milosgajdos** [comment]
Hi, one of the maintainers of `distribution/distribution`.

You are right to assume `distribution/distribution` is the CNCF project. It was donated by Docker to CNCF a few years ago. Again, you correctly assumed that there is a relation b/w the official `registry` image and the one in the `distribution` namespace in Docker Hub.

At the moment the official images are built of the `2.x` release branch fo `dist/dist` GH repository. This branch has been put into maintenance mode (releases are very rare and mostly related to security updates) until the `v3` release is cut (no ETA). That being said, any `v2.x` releases are also released as new Docker Official Images (DOI).

What you see in Docker Hub in `distribution` namespace, particularly the `edge` tag is what gets pushed to it automatically every time something is merged into `main` branch. This is a temporary solution until `v3` is cut and is considered to be unstable.

We, the maintainers, need to figure out the path forward with regard to DOI pipeline once the `v3` release arrives.

**Comment by thetredev** [comment]
@milosgajdos thanks for explaining!

**Comment by tianon** [comment]
This can probably be considered resolved now that the GitHub repository has moved? :eyes:

**Comment by thaJeztah** [comment]
Looks like the readme could use some TLC to scrub some of the Docker references, links to docker's docs, and update it to refer to the OCI distribution spec (not Docker Registry v2 spec)

![IMG_0166](https://github.com/distribution/distribution-library-image/assets/1804568/54b70912-90c7-46ff-a6ae-3d199a6b90fb)
![IMG_0165](https://github.com/distribution/distribution-library-image/assets/1804568/7d1ace2b-a85b-4602-aa99-5020c7341496)
![IMG_0167](https://github.com/distribution/distribution-library-image/assets/1804568/4464e4ba-1db8-4b62-b8c7-5c84e37361bb)

**Comment by milosgajdos** [comment]
How is this README generated? Distribution maintainers maintain the readme for https://hub.docker.com/r/distribution/distribution, but I'm not sure how the DOI readme is done

**Comment by milosgajdos** [comment]
Docs update PR here: https://github.com/docker-library/docs/pull/2402

Once that's merged in, we can close this issue 😄

**Comment by thaJeztah** [comment]
OHMAN; sorry for the _huge_ screenshots; posted them from my phone, and I keep forgetting those hi-dpi ones get posted so large 😂 

I'll look at your PR; thanks!

**Comment by milosgajdos** [comment]
Closing as https://github.com/docker-library/docs/pull/2402 has been merged
